Improve perf for iterating over tracked entities

ajcvickers opened the pull request:
Part of #8898
Fixes #14231

Investigation into DbSet.Local shows that:
* Iteration over tracked entities had a lot of LINQ code; fixed by un-LINQing
* Multiple composed iterators cause slow-down; fixed by giving LocalView it's only IStateManager method
* Filtering causes slow-down; fixed by splitting storage into multiple dictionaries. Makes lookup for entity instance slightly slower.
* Calculate Count lazily

This means that DbSet.Local iteration is now about 3x faster, although it depends a lot on what is being tracked. Getting an ObservableCollection and then iterating over that _once_ is slower than just iterating over the entries once. Iterating over the ObservableCollection multiple times is still faster than iterating over DbSet.Local multiple times. This seems like a reasonable trade-off. If the application does heavy iteration, then creating a new collection for that seems reasonable.

With regard to #8898, the layering we have still seems correct, so moving forward with the breaking changes there.

AndriySvyryd commented:
Do you have an idea of how much slower on average is it to get a single entry now?

AndriySvyryd commented on src/EFCore/ChangeTracking/Internal/EntityReferenceMap.cs:
```diff
@@ -0,0 +1,351 @@
+            {
+                foreach (var entry in _addedReferenceMap.Values)
+                {
+                    yield return entry;
```
Might be faster to do an explicit `Concat` instead

ajcvickers replied:
Didn't seem to be, but I could test it again.

AndriySvyryd replied:
Also you can check whether the map is empty before iterating

ajcvickers replied:
I knew I shouldn't have done any perf work...

ajcvickers replied:
Too hard to get anything signed-off. :trollface:

ajcvickers replied:
Concat was significantly slower. Checking for empty was barely measurable, but doing it anyway, partly because there is then a further optimization which makes the iteration faster if we only need to return entries for one state--by not introducing another iterator state machine in this case.

ajcvickers commented:
@AndriySvyryd I don't have numbers, but I could do a quick test for it in a tight loop and see. But then, is a test with a tight loop really valid?

AndriySvyryd commented on src/EFCore/ChangeTracking/Internal/IStateManager.cs:
```diff
@@ -95,6 +95,23 @@ public interface IStateManager : IResettableService
+            bool added = false,
+            bool modified = false,
+            bool deleted = false,
+            bool unchanged = false);
```
I don't like that without any arguments this returns an empty enumerable, I'd prefer a parameterless overload and a no-defaults overload instead.

ajcvickers replied:
The "parameterless overload" is just `Entries`. Maybe this is just a naming thing? How about `GetEntriesForStates()`?

AndriySvyryd replied:
No, I'm more concerned about the default values, but this isn't critical unless somebody else agrees with me. @divega @bricelam

ajcvickers replied:
It looks very nice from the consuming side:
```C#
GetEntries(unchanged: true, modified: true);
```

ajcvickers replied:
Also, this is internal API. :-)

ajcvickers replied:
Changed it's name; I like it; no one else commented; it's internal. :-)

AndriySvyryd commented on src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:
```diff
@@ -1725,12 +1725,9 @@ protected virtual IEnumerable<MigrationOperation> GetDataOperations()
             if (_sourceStateManager != null)
             {
-                foreach (var sourceEntry in _sourceStateManager.Entries.ToList())
+                foreach (var sourceEntry in _sourceStateManager.GetEntries(added: true).ToList())
```
Another optimization you could try it to implement `Count` in the state manager with the same parameters as `GetEntries` and use it to preallocate the list of the exact size you will need to avoid resizing.

ajcvickers replied:
Hmm. I did part of this and it didn't make much difference, but I like the idea of pushing the Count logic to the state manager. Will look into that.

ajcvickers replied:
Did something like this and added extension methods for creating the list of the correct capacity, populating, and returning.

AndriySvyryd commented:
@ajcvickers It's valid to give us an idea. Or you could run the existing micro-benchmark

ajcvickers commented:
@AndriySvyryd For `TryGetEntry` in a tight loop on my machine, for 50,000,000 iterations:
* Unchanged: no measurable difference (~1800ms)
* Modified: about 28% slower (~2300ms)
* Added: about 56% slower (~2800ms)
* Deleted: about 100% slower (~3600ms)

AndriySvyryd commented:
@ajcvickers Ok, that's not so bad

ajcvickers commented:
@AndriySvyryd Do you think the order that we check the dictionaries is optimal? That is, should it be something other than `Unchanged` -> `Modified` -> `Added` -> `Deleted`?

AndriySvyryd commented:
@ajcvickers We might flip `Modified` and `Added`, but I don't have a strong argument one way or the other

AndriySvyryd commented on src/EFCore/ChangeTracking/Internal/EntityReferenceMap.cs:
```diff
@@ -0,0 +1,474 @@
+
+            if (!hasDependentTypes)
+            {
+                var count
```
`numberOfStates`?
